src/react/controllers/viewport-controller.js

```diff
@@ -0,0 +1,132 @@
+
+  /** Viewport constraints */
+  // Max zoom level
+  maxZoom: PropTypes.number,
+  // Min zoom level
+  minZoom: PropTypes.number,
+  // Max pitch in degrees
+  maxPitch: PropTypes.number,
+  // Min pitch in degrees
+  minPitch: PropTypes.number,
+
+  /**
+   * `onViewportChange` callback is fired when the user interacted with the
+   * map. The object passed to the callback contains viewport properties
+   * such as `longitude`, `latitude`, `zoom` etc.
+   */
+  onViewportChange: PropTypes.func,
+
+  /** Enables control event handling */
+  // Scroll to zoom
+  scrollZoom: PropTypes.bool,
+  // Drag to pan
+  dragPan: PropTypes.bool,
+  // Drag to rotate
+  dragRotate: PropTypes.bool,
+  // Double click to zoom
+  doubleClickZoom: PropTypes.bool,
+  // Pinch to zoom / rotate
+  touchZoomRotate: PropTypes.bool,
+
+  /** Accessor that returns a cursor style to show interactive state */
+  getCursor: PropTypes.func,
+
+  // A map control instance to replace the default map controls
+  // The object must expose one property: `events` as an array of subscribed
+  // event names; and two methods: `setState(state)` and `handle(event)`
+  controls: PropTypes.shape({
+    events: PropTypes.arrayOf(PropTypes.string),
+    handleEvent: PropTypes.func
+  })
+};
+
+const getDefaultCursor = ({isDragging}) => isDragging ? CURSOR.GRABBING : CURSOR.GRAB;
+
+const defaultProps = {
+  onViewportChange: null,
+
+  scrollZoom: true,
+  dragPan: true,
+  dragRotate: true,
+  doubleClickZoom: true,
+  touchZoomRotate: true,
+
+  getCursor: getDefaultCursor
+};
+
+export default class Controller extends PureComponent {
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      isDragging: false      // Whether the cursor is down
+    };
+  }
+
+  componentDidMount() {
+    const {eventCanvas} = this.refs;
+
+    this._eventManager = new EventManager(eventCanvas);
+
+    // If props.controls is not provided, fallback to default MapControls instance
+    // Cannot use defaultProps here because it needs to be per map instance
+    this._controls = this.props.controls || new Controls(this.props.StateClass);
+
+    this._controls.setOptions(Object.assign({}, this.props, {
+      onStateChange: this._onInteractiveStateChange.bind(this),
+      eventManager: this._eventManager
+    }));
+  }
+
+  componentWillUpdate(nextProps) {
+    if (this._controls) {
+      this._controls.setOptions(nextProps);
+    }
+  }
+
+  componentWillUnmount() {
+    this._eventManager.destroy();
+  }
+
+  _onInteractiveStateChange({isDragging = false}) {
+    if (isDragging !== this.state.isDragging) {
+      this.setState({isDragging});
+    }
+  }
+
+  render() {
+    const {width, height, getCursor} = this.props;
+
+    const eventCanvasStyle = {
+      width,
+      height,
+      position: 'relative',
+      cursor: getCursor(this.state)
+    };
+
+    return (
+      createElement('div', {
+        key: 'map-controls',
+        ref: 'eventCanvas',
```

Comment:
String refs are being removed in React 16. See [this thread on react-map-gl](https://github.com/uber/react-map-gl/pull/316).

Reply:
> String refs are being removed in React 16. See this thread on react-map-gl.

OK. We are using them in more places, maybe just open a separate PR to remove them all?

Reply:
Opened https://github.com/uber/deck.gl/issues/917